### ISSUE TITLE
confusing performance_metrics 

### ISSUE THREAD

**Issue by ggrrll** [facebook/prophet#1773]
Hi, 

I guess I am having the same issues as others ( #941 ?) to understand `performance_metrics`

What one would typically expect to have when doing **k-CV** is k values , for each validation metric.
In the case of prophet, if I understood correctly, **k** = # of cutoffs 
 (which should be also   ` k = int( (df.shape[0] - initial - horizon) / period) + 1` )
(just as double check - can you confirm?)

* how can I get a `df_p` (performance_metrics) whose size = # cutoffs? 

* the confusion arises also from the fact that you are probably using term `horizon` with two different meanings:
in [the docs](https://facebook.github.io/prophet/docs/diagnostics.html#hyperparameter-tuning) at the beginning it seems to be a **fixed** window size, while after, it becomes a column with many values... 

Thanks

**Comment by bletham** [comment]
I can see how the overloaded use of `horizon` can be confusing here. Hopefully this will help and then we'll have to see if there is a way this can be made more clear in the documentation.

First on the number of cutoffs: In a normal k-fold CV you would generate k train/test sets by partitioning the data, from which you get k estimates of prediction error. The idea here is basically the same, except because we have to maintain temporal structure we don't just randomly partition the data like in a usual k-fold CV. We select k cutoffs, which produces k train/test set. The train set is all of the time series before that cutoff, the test set is the window of length `horizon` after the cutoff. As in usual k-fold CV, we can evaluate the model on each of these train/test sets to get k estimates of prediction error.

The formula you give for the number of cutoffs is basically correct; things can get complicated if the time series frequency is not regular, in which case if the cutoffs are regular you might end up with an empty test set so there is some logic to avoid that that might reduce the number of cutoffs:
https://github.com/facebook/prophet/blob/20f590b7263b540eb5e7a116e03360066c58de4d/python/fbprophet/diagnostics.py#L41-L43
but that's really an edge case, and in the typical situation the formula you give is correct.

Now for the meaning of `horizon`. The `horizon` that is input to `cross_validation` is the length of the window for which forecasts will be made (that is, the length of the test set), as illustrated in the figure at the top of https://facebook.github.io/prophet/docs/diagnostics.html. The `horizon` in the output of `performance_metrics` is the actual horizon of that particular prediction. Meaning: if you look at the output of `cross_validation` you will see that it contains all of the test-set predictions. For every test-set point, we see the `ds` for that point, the `cutoff` for the model that was used to predict that point, the true `y` value, and the model predicted `yhat`. The horizon for that particular prediction is how many days into the future that prediction was: `ds - cutoff`. So the test set will actually contain forecasts at a range of horizons: if you have daily data and set the test-set window horizon to 30 days, then you will actually be making 30 predictions with horizons ranging from 1 to 30 days (each date in the test set). I hope that makes sense.

Now let's get to `performance_metrics`. The `performance_metrics` function computes various error metrics and averages them. In a regular k-fold CV, there are two distinct averages: an average over each test-set prediction, and an average over folds. The points in the test-set are all exchangeable so there is no reason to not throw them all together into a single average A similar thing happens here where we can average over both folds and over the different points in the test-set, but it's a little different now because in a time series the test-set points are not exchangeable and so we might not want to just average over them all. Recall that the test set contains predictions for a variety of distances into the future: Some of those test-set predictions may be easier than others. I might have lower error when predicting 1 day after the cutoff than I have for the prediction that is 30 days after the cutoff. Moreover, I might be interested in understanding how my prediction error grows as I look further into the future.

The `performance_metrics` function always averages over the k folds (cutoffs), but what it does for the points within the test set is a little more complicated. In order to measure how error depends on the distance into the future being forecasted, it does a rolling average across the test set. This is what is being illustrated in the figure in https://facebook.github.io/prophet/docs/diagnostics.html for `plot_cross_validation_metric`: the gray points are all of the errors for each individual predictions (so, a dot for every row in `df_cv`, plot against `df_cv['ds'] - df_cv['cutoff']`), and the blue line is the average over cutoffs and the rolling average over "horizon", where here "horizon" is `df_cv['ds'] - df_cv['cutoff']`, the actual horizon for that test-set point.

If you do `performance_metrics(df_cv, rolling_window=0)` then it will average over folds but not over time; so with 30 day horizon and daily data, you would have 30 rows, each of which is the average error across folds for points that were X days after the cutoff, where X is what is given in the `horizon` column of the output.

If you do `performance_metrics(df_cv, rolling_window=1)`, it would be more like a regular CV where the average is taken over all of the test-set points, so there will be a single row. This estimates the average forecast error over the entire 30-day horizon.

The `performance_metrics` function always averages over cutoffs. It sounds like you are interested in a different approach where you get a single average _per cutoff_. So, you average over the entire test set but don't average over cutoffs. This would make sense for instance for seeing what the variance is across cutoffs. The `performance_metrics` function won't be helpful for this task, you'll have to compute the error yourself. But everything you need is in the output of `cross_validation`, call it `df_cv`, so it shouldn't be too challenging. Basically you would just group by `cutoff` and then compute average error (for instance, MSE is the average of `(df_cv['yhat'] - df_cv['y']) ** 2`) across all of the points with the same `cutoff`.

**Comment by ggrrll** [comment]
Thanks for this very thorough answer !
here, you totally got my point:
>  This would make sense for instance for seeing what the variance is across cutoffs. 

  -- you have basically re-written the docs :), so you can maybe integrate it in the docs

**Comment by ivangvi** [comment]
Hi guys, thanks for the interesting discussion. I was wondering, isn't it a problem that the CV procedure currently takes an expanding training set? For instance, if we have:

- 365 days of data.
- 180 set as initial.
- 1 set as period and horizon.

Each fold will incrementally have 1 more day of data, up to the last one having 364 to train versus the first one having 182. In my experience in forecasting either with Prophet or LightGBM, the amount of days for training can have a considerable impact on actual, out of sample performance. Therefore, the current CV procedure will be averaging folds with different amount of training data and will likely be less faithful to actual out of sample performance. Sometimes I found taking less data can be better to capture recent patterns and don't even look into old ones.

Maybe the procedure was thought for scenarios where it's always better to expand data as it comes and not have a fixed rolling window for training? 

Thanks for your insights, regards.

**Comment by ggrrll** [comment]
so, indeed, CV in time series is more tricky ...

here is a nice post describing the different possible approaches ( I guess the main ones -- I am not a big expert)
https://medium.com/@soumyachess1496/cross-validation-in-time-series-566ae4981ce4

I guess the devs here (@bletham , ...) could think of integrating few other in the futurue (like the _Blocked CV_, which is what @ivangvi  basically proposed, if  I understood...)

**Comment by ofwallaart** [comment]
Very interesting topic and discussion above! Would indeed love to see some other CV techniques implemented. @bletham is there a specific reason why you chose to implement Time Series Split Cross-Validation rather then Blocked Cross-Validation?

**Comment by bletham** [comment]
An advantage of the split CV is that it allows you to get CV estimates with longer histories. Given a time series of length T and horizon H, the goal of CV is to estimate how well the forecast will predict out H steps. Let's say that I want 10 estimates of error (10 folds). With a block design, I would break the time series into 10 chunks, each of which thus has length T/10, and I train on them. This means I'm using _very_ short time series in my CV. The forecast error from a very short time series probably is not a good indicator of the forecast error for my full time series, which means the CV is pointless. That is, if I have 10x more data do I expect to be able to do a better job forecasting? If so, then the block CV is not providing a good estimate.

Imagine T is 30 months and H is 3 months; that is, I have 2.5 years of data and I'm trying to predict 3 months, a very reasonable task. To get 5 folds (not even very many!) would mean each CV block is 6 months, predicting out 3 months. Will this provide a reasonable error estimate for the full time series that is fit to 30 months of data? Most likely not. For instance, if I have yearly seasonality, I will not be able to estimate at all the yearly seasonality within the CV window, since it is only 6 months long. Does it make sense to measure error without modeling yearly seasonality if my full model does include yearly seasonality? The CV model is so fundamentally different from the full model that it just doesn't provide a meaningful error metric.

With the split CV approach, I would construct the 5 folds by putting cutoffs at 27, 24, 21, 18, and 15 months. All of these are longer than 12 months, so all of them would be able to fit yearly seasonality. They provide correlated estimates of prediction error, but they are much better estimates because the CV subproblem (e.g. "given 24 months, predict 3") is much closer to the real problem I'm trying to estimate the error for ("given 30 months, predict 3").

**Comment by ivangvi** [comment]
Hi all, good contributions. Just to clarify:

1. Blocked CV:
![image](https://user-images.githubusercontent.com/30244899/107580527-73cdb800-6bd5-11eb-9b01-99c9e3b2a4ed.png)
2. Rolling/Sliding CV:
![image](https://user-images.githubusercontent.com/30244899/107580705-b8595380-6bd5-11eb-9b13-6985bcf7e191.png)
3. Expanding CV:
![image](https://user-images.githubusercontent.com/30244899/107580722-c0b18e80-6bd5-11eb-9cb9-1b3be8929da1.png)

In my original post I was talking about 2, not 1. I agree with you @bletham, blocked is very punishing for scarce data, I haven't used it and can't think of any advantage over rolling.

I had a use case where I had a lot of data and even looking too much into the past was detrimental for the model, hence the sliding approach in order to be able to parametrize how much past data to get. For this case, built-in sliding option would have helped.

Of course, I still think the expanding default is great for any case where one detects that the more data the model gets, the better the prediction, regardless of data/pattern drift. It is clear that this is the best approach for whenever one would then proceed to use all data for training. Nevertheless, on a production pipeline, I'd still closely monitor performance to detect any breakpoint when looking too much into the past starts to be detrimental.

2/3 images reference: https://stats.stackexchange.com/questions/326228/cross-validation-with-time-series

Thanks!

**Comment by bletham** [comment]
Yeah I don't think there is necessarily a single correct answer for all time series. Having training sets that don't overlap with eachother and test sets that don't overlap with eachother means the estimates of error will be statistically independent, which is useful for considering the variance across CV folds. The downside is that they may be biased because they are using quite a bit less data than the full train will use (unless you intend after CV to train using only the last chunk of data, in which case it is definitely what you'd want to do). The relative values of using longer training sets vs. using independent training sets will really depend on the time series.

**Comment by ofwallaart** [comment]
Great answers guys! Much appreciated, as often in this field of science the answer is: it depends. Very interesting to read about your considerations and trade-offs.
